Source/Core/ProofObligationDescription.cs

```diff
@@ -0,0 +1,104 @@
+namespace Microsoft.Boogie;
+
+/// <summary>
+/// A multi-faceted description of a proof obligation. This class is intended
+/// to provide several forms of human-readable text summarizing the meaning of
+/// a proof obligation: one for the case where the proof has succeeded, one
+/// for the case where the proof has failed, and one for the case of quickly
+/// identifying the type of proof obligation (for inclusion in long lists of
+/// proof obligations, for example).
+/// </summary>
+public abstract class ProofObligationDescription {
+  /// <summary>
+  /// A description of what this proof obligation means when it has been
+  /// successfully proven.
+  /// </summary>
+  public abstract string SuccessDescription { get; }
+
+  /// <summary>
+  /// A description of what this proof obligation means (or might mean)
+  /// when the prover has failed to establish its validity.
+  /// </summary>
+  public virtual string FailureDescription =>
+    $"Failed to prove: {SuccessDescription}";
+
+  /// <summary>
+  /// A short description of the general type of this proof obligation.
+  /// </summary>
+  public abstract string ShortDescription { get; }
+}
+
+public class AssertionDescription : ProofObligationDescription
+{
+  public override string SuccessDescription => "This assertion holds.";
+
+  public override string FailureDescription => "This assertion might not hold.";
+
+  public override string ShortDescription => "assert";
+}
+
+public class PreconditionDescription : ProofObligationDescription
+{
+  public override string SuccessDescription =>
+    "All preconditions hold for call.";
+
+  public override string FailureDescription =>
+    "A precondition for this call might not hold.";
+
+  public override string ShortDescription => "precondition";
+}
+
+public class RequiresDescription : ProofObligationDescription
```

Comment:
What is the difference between "preconditions" and "requires"? I'm wondering if we're using mostly the same messages for both because Dafny doesn't distinguish, but there IS a distinction at the Boogie level?

Reply:
The way these messages are used, "precondition" refers to the set of all things that must be proved to call a procedure whereas "requires" refers to a specific requires clause. So a precondition is basically a set of requires clauses. Boogie's error messages use both descriptions. Dafny's might not. But to allow this PR to leave all of Boogie's output unchanged it needs to have classes that distinguish the concepts.